### DISCUSSION TITLE
How to debug javascript issues in formula?

### DISCUSSION THREAD

**Discussion by leolivier** [Budibase/budibase#12670] [Help] [answered]
Hi, 
I wrote the following javascript formula for computing the age of a person:
```
const birthDate = new Date($("birthdate"));
const nowDate = Date.now();

let years = (nowDate.getFullYear() - birthDate.getFullYear());

if (nowDate.getMonth() < birthDate.getMonth() || 
  nowDate.getMonth() == birthDate.getMonth() && 
  nowDate.getDate() < birthDate.getDate()) {
  years--;
}

return years;
```
I did put that code in a dynamic formula in a table:
![image](https://github.com/Budibase/budibase/assets/4014752/8fcb0a1b-68a1-4e43-80f2-b21f6ca16017)

but when I create a row in the table, I get an error like this:
![image](https://github.com/Budibase/budibase/assets/4014752/fff12163-9913-4232-aa0a-a8ef06170247)

How can I know what is the error?

**Comment by leolivier** [comment]
I debugged the JS code using other means but still the question remains: how to get the JS errors in Budibase?

**Reply by ConorWebb96** [reply, marked as answer]
Hey @leolivier,

Currently you can't debug JS within Budibase, it requires a lot of guess work sometimes in order to get your JS fixed and working. We do have an internal ticket to improve upon this, so errors are more easily highlighted. 

However, its currently not on the road map. My approach when writing complicated JS is to use Codepen, SourceLair, jsFiddle etc. for testing then add it into the area i wish to use within Budibase.

Hopefully this makes sense.

**Comment by leolivier** [comment]
Thanks @ConorWebb96 
Actually, I do the same (I'm using playcode.io), but often, the issue is not so simple, e.g.:
* it can be in the bindings (hard to guess if your $("xxx") expression is working properly or not) 
* it can lie in the accepted or not predefined JS: e.g. for guessing the current language, I used navigator.language in the frontend but it looks like it's executed on the backend and it doesn't work althoough it works perfectly well in playcode.io...
I think the best way to manage this would be to allow testing directly the result in the javascript editor of budibase (could be interesting for handlebars too btw)
